File: app/backend/app/routers/stats.py
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.deps import get_db
from app.schemas.stats import (
    ComparisonStats,
    LeaveItem,
    RangeStats,
    StudentStatsRow,
    TodayStats,
)
from app.services import stats_service

router = APIRouter(prefix="/stats", tags=["stats"])


@router.get("/today", response_model=TodayStats)
def stats_today(db: Session = Depends(get_db)):
    return stats_service.today_summary(db)


@router.get("/range", response_model=RangeStats)
def stats_range(
    from_date: date = Query(..., alias="from"),
    to_date: date = Query(..., alias="to"),
    granularity: str = Query("day", pattern="^(day|week|month)$"),
    db: Session = Depends(get_db),
):
    return stats_service.range_stats(
        db, from_date=from_date, to_date=to_date, granularity=granularity
    )


@router.get("/students", response_model=list[StudentStatsRow])
def stats_students(
    from_date: date = Query(..., alias="from"),
    to_date: date = Query(..., alias="to"),
    db: Session = Depends(get_db),
):
    return stats_service.student_ranking(db, from_date=from_date, to_date=to_date)


@router.get("/leave", response_model=list[LeaveItem])
def stats_leave(
    from_date: date = Query(..., alias="from"),
    to_date: date = Query(..., alias="to"),
    db: Session = Depends(get_db),
):
    return stats_service.leave_list(db, from_date=from_date, to_date=to_date)


@router.get("/comparison", response_model=ComparisonStats)
def stats_comparison(
    period: str = Query("week", pattern="^(day|week|month)$"),
    from_date: date | None = Query(None, alias="from"),
    to_date: date | None = Query(None, alias="to"),
    db: Session = Depends(get_db),
):
    if (from_date is None) != (to_date is None):
        raise HTTPException(status_code=422, detail="from 和 to 必须同时提供")
    if from_date is not None and to_date is not None and from_date > to_date:
        raise HTTPException(status_code=422, detail="from 不能晚于 to")
    return stats_service.comparison(
        db,
        period=period,
        from_date=from_date,
        to_date=to_date,
    )

File: app/frontend/src/api/stats.ts
import { api } from './client'
import type {
  ComparisonStats,
  LeaveItem,
  RangeStats,
  StudentStatsRow,
  TodayStats,
} from './types'

function comparison(period: 'week' | 'month'): Promise<ComparisonStats>
function comparison(
  from: string,
  to: string,
  period: 'day' | 'week' | 'month',
): Promise<ComparisonStats>
function comparison(
  fromOrPeriod: string,
  to?: string,
  period?: 'day' | 'week' | 'month',
) {
  const params = to && period
    ? { from: fromOrPeriod, to, period }
    : { period: fromOrPeriod }
  return api
    .get<ComparisonStats>('/stats/comparison', { params })
    .then((response) => response.data)
}

export const statsApi = {
  today: () => api.get<TodayStats>('/stats/today').then((r) => r.data),

  range: (from: string, to: string, granularity: 'day' | 'week' | 'month' = 'day') =>
    api
      .get<RangeStats>('/stats/range', { params: { from, to, granularity } })
      .then((r) => r.data),

  students: (from: string, to: string) =>
    api
      .get<StudentStatsRow[]>('/stats/students', { params: { from, to } })
      .then((r) => r.data),

  leave: (from: string, to: string) =>
    api.get<LeaveItem[]>('/stats/leave', { params: { from, to } }).then((r) => r.data),

  comparison,
}
